tests/wallet/cat_wallet/test_cat_wallet.py

```diff
@@ -827,3 +838,162 @@ async def check_wallets(node):
 
         await time_out_assert(20, cat_wallet.get_confirmed_balance, 35)
         await time_out_assert(20, cat_wallet.get_unconfirmed_balance, 35)
+
+    @pytest.mark.parametrize(
+        "trusted",
+        [True, False],
+    )
+    @pytest.mark.asyncio
+    async def test_cat_change_detection(self, self_hostname: str, two_wallet_nodes_services: Any, trusted: Any) -> None:
+        num_blocks = 1
+        full_nodes, wallets, bt = two_wallet_nodes_services
+        full_node_api: FullNodeSimulator = full_nodes[0]._api
+        full_node_server = full_node_api.full_node.server
+        wallet_service_0 = wallets[0]
+        wallet_node_0 = wallet_service_0._node
+        wallet_node_1 = wallet_service_0._node
+        wallet_0 = wallet_node_0.wallet_state_manager.main_wallet
+        wallet_1 = wallet_node_1.wallet_state_manager.main_wallet
+
+        client_0 = await WalletRpcClient.create(
+            bt.config["self_hostname"],
+            wallet_service_0.rpc_server.listen_port,
+            wallet_service_0.root_path,
+            wallet_service_0.config,
+        )
+        wallet_node_0.config["automatically_add_unknown_cats"] = True
+
+        if trusted:
+            wallet_node_0.config["trusted_peers"] = {
+                full_node_api.full_node.server.node_id.hex(): full_node_api.full_node.server.node_id.hex()
+            }
+        else:
+            wallet_node_0.config["trusted_peers"] = {}
+        wallet_node_0.config["automatically_add_unknown_cats"] = True
+
+        await wallet_node_0.server.start_client(PeerInfo(self_hostname, uint16(full_node_server._port)), None)
+        await full_node_api.farm_blocks_to_wallet(count=num_blocks, wallet=wallet_0)
+        await full_node_api.wait_for_wallet_synced(wallet_node=wallet_node_0, timeout=20)
+
+        # Mint CAT to ourselves, immediately spend it to an unhinted puzzle hash that we have manually added to the DB
+        # We should pick up this coin as balance even though it is unhinted because it is "change"
+        intermediate_sk_un = master_sk_to_wallet_sk_unhardened_intermediate(
+            wallet_node_0.wallet_state_manager.private_key
+        )
+        pubkey_unhardened = _derive_path_unhardened(intermediate_sk_un, [100000000]).get_g1()
+        inner_puzhash = puzzle_hash_for_pk(pubkey_unhardened)
+        puzzlehash_unhardened = construct_cat_puzzle(
+            CAT_MOD,
+            Program.to(None).get_tree_hash(),
+            inner_puzhash,  # type: ignore
+        ).get_tree_hash_precalc(inner_puzhash)
+        change_derivation = DerivationRecord(
+            uint32(0),
+            puzzlehash_unhardened,
+            pubkey_unhardened,
+            WalletType.CAT,
+            uint32(2),
+            False,
+        )
+        # Insert the derivation record before the wallet exists so that it is not subscribed to
```

Comment:
> I just wondered because you anyway derive something which is far beyond the range of what the wallet will derive on wallet creation and with that wouldn't subscribe to it, no?

Yeah but if I insert it after wallet creation the wallet will start deriving indexes _after_ the new one I insert for new transactions and then the farming new coins method on the simulator and it's just a whole mess. It's just simpler to put this in first before the wallet is created.

> Why so? From my understanding its especially here to make the other PR possible, like making it backwards compatible to all existing CAT change outputs out there which aren't hinted so the test will still be relevant after that PR unless i misunderstand something

Ah yeah you're right, it _would_ be moot if https://github.com/Chia-Network/chia-blockchain/pull/15179 didn't exist, but that's going to stop creating the scenario that this is trying to fix in our local tests.  So yeah you're right, ignore my comment to that effect.